### PR TITLE
fix/reduced infura consumption with websocket connection

### PR DESCRIPTION
**Pull request by wpenn**
**Before submitting this PR, please make sure**:
- [ ] Your code builds clean without any errors or warnings
- [ ] You are using approved title ("feat/", "fix/", "docs/", "refactor/", etc)

**A description of the changes proposed in the pull request**:

Reduced infura consumption below 100,000 per day limit with WebSocket connection to replace watcher for new blocks
- Clubhouse Ticket: https://app.clubhouse.io/coinalpha/story/9071/reduce-infura-api-calls-on-hb-client
- Github: #1848 

Need someone to also to test if the consumption is indeed below the rate (4166/hr) for them

### CODE REVIEW

**Comment by dennisocana** [comment]
Ran a bot overnight with 3 orders per side. Total method calls are at 39,856 for 10 hours.

image
image

I will stop the instance for now since I ran into errors due to Radar Relay's request limit.

[logs_rr_weth-dai.log](https://github.com/CoinAlpha/hummingbot/files/4887617/logs_rr_weth-dai.log)

**Comment by dennisocana** [comment]
OK to test

File: hummingbot/templates/conf_global_TEMPLATE.yml [hunk 1]
```diff
@@ -45,6 +45,7 @@ celo_password: null
 # Ethereum wallet address: required for trading on a DEX
 ethereum_wallet: null
 ethereum_rpc_url: null
+ethereum_rpc_ws_url: null
```
Comment (Nullably) [inline hummingbot/templates/conf_global_TEMPLATE.yml:48]: We need to increase the template_version by one every time we add a new param.

File: hummingbot/wallet/ethereum/watcher/websocket_watcher.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+                        if incoming_block is not None:
+                            new_block: AttributeDict = await self.call_async(self._w3.eth.getBlock,
+                                                                             incoming_block.get("hash"), True)
+                            self._current_block_number = new_block.get("number")
+                            self._block_cache[new_block.get("hash")] = new_block
+                            self.trigger_event(NewBlocksWatcherEvent.NewBlocks, [new_block])
+                except asyncio.CancelledError:
+                    if self._network_on:
+                        self.logger().network("Network closed abnormally, reconnecting ...")
+                        self.connect()
```
Comment (Nullably) [inline hummingbot/wallet/ethereum/watcher/websocket_watcher.py:140]: Not sure if we should `connect()` here on this exception, seems to me the task is cancelled and exit from here. To connect here might be unnecessary.

File: hummingbot/wallet/ethereum/watcher/websocket_watcher.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        nonce = await self._send(emit_data)
+        raw_message = await self._client.recv()
+        if raw_message is not None:
+            resp = ujson.loads(raw_message)
+            if resp.get("id", None) == nonce:
+                self._node_address = resp.get("result")
+                return True
+        return False
+
+    async def unsubscribe(self, params) -> bool:
```
Comment (Nullably) [inline hummingbot/wallet/ethereum/watcher/websocket_watcher.py:109]: Do we need this? Is it used anywhere?

File: hummingbot/wallet/ethereum/watcher/websocket_watcher.py (hunk 2 above)
Comment (wpenn) [reply hummingbot/wallet/ethereum/watcher/websocket_watcher.py:140]: The Websocket sometimes "abnormally" disconnects, for example when my computer sleeps with the bot running; this accounts for the "abnormal disconnections" when there's a CancelledError but the WebSocket is still intended to be connected; this replaces the self._client with a connected WebSocket. Let me know if I should figure out a workaround or just remove it

File: hummingbot/wallet/ethereum/watcher/websocket_watcher.py (hunk 3 above)
Comment (wpenn) [reply hummingbot/wallet/ethereum/watcher/websocket_watcher.py:109]: It's not used, but would be necessary if this became generalized for handling WebSockets. I can remove it

File: hummingbot/wallet/ethereum/watcher/websocket_watcher.py (hunk 2 above)
Comment (Nullably) [reply hummingbot/wallet/ethereum/watcher/websocket_watcher.py:140]: Does the `CancelledError` raise on those situations, like when the websocket abnormally disconnects? From what I understand the CancelledError is raised only when `cancel()` is called, and it is a way to wrap up and finish the task. If you need to reconnect, it's fine but I don't think this is the right place for it.

File: hummingbot/wallet/ethereum/watcher/websocket_watcher.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+                                                          "Check wallet network connection")
+                except BlockNotFound:
+                    pass
+                except ConnectionClosedOK:
+                    raise
+                except Exception as e:
+                    self.logger().network(f"Error fetching new block: {e}", exc_info=True,
+                                          app_warning_msg="Error fetching new block. "
+                                                          "Check wallet network connection")
+        except asyncio.CancelledError:
```
Comment (Nullably) [inline hummingbot/wallet/ethereum/watcher/websocket_watcher.py:154]: This except and raise here is unnecessary, you need to do this only when you also catch other kinds of Exceptions.

File: hummingbot/templates/conf_global_TEMPLATE.yml (hunk 1 above)
Comment (Nullably) [reply hummingbot/templates/conf_global_TEMPLATE.yml:48]: What I mean is you need to update `template_version` (line 6 of this file) to 9.

File: hummingbot/wallet/ethereum/watcher/websocket_watcher.py (hunk 2 above)
Comment (wpenn) [reply hummingbot/wallet/ethereum/watcher/websocket_watcher.py:140]: Made a mistake. Wrong location; I put it in ConnectionClosedError from WebSockets instead. This is for when connectionclosed errors codes are not 1000 (OK) or 1001 ("Going Away"). Trying to figure out if there's a way to check connection another way in the fetch loop

Comment (wpenn) [reply hummingbot/wallet/ethereum/watcher/websocket_watcher.py]: @Nullably pushed up changes, what do you think?

File: hummingbot/wallet/ethereum/watcher/websocket_watcher.py (hunk 4 above)
Comment (Nullably) [reply hummingbot/wallet/ethereum/watcher/websocket_watcher.py:143]: You don't need to except asyncio.CancelledError here, but you do need one in your inner try-except block.

**Pull request merged**